### PR TITLE
Session management + user properties + client to send events

### PR DESCRIPTION
**Pull request by eliasyishak**
This PR contains all the action items from issue: https://github.com/eliasyishak/dash_analytics/issues/14

A session management class has been created that is responsible for extending the session's `last_ping` variable if we are within the session duration – if it is not, we will create a new session id.

A user properties class has also been added to make it easier to pass information about the user and their machine. It also comes with its own methods that will prepare the data it contains for the json payload that will get sent to the GA4 endpoint for data collection.

A GA client class has also been added to handle sending the POST request itself with the json payload for data collection

-------

General updates for this PR:

`analytics.dart`
- added platform as an arg for the class
- using the `constants.dart` file to store constant values that are getting passed 
- added a close method which will close the http client in `GAClient`
- added the sendEvent method to take in `eventName` and `eventData` and use `GAClient` to send data
- instead of default values in the `AnalyticsImpl` class, we will use the factory constructor in the abstract class to add the default values (allows for easier testing)
- using the `GAClient` class within the `AnalyticsImpl` class to send data
- creating a `UserProperty` instance in the constructor for `AnalyticsImpl` to store metadata around the user

`config_handler.dart`
- using `package:clock` for date stamps which allows for easier testing when checking session length
- using the `constants.dart` file to store the directory and file names for the files generated

`ga_client.dart`
- created this class to send events to Google Analytics to abstract away creating the `HttpClient` from the main `Analytics` class

`initializer.dart`
- moving the functions that created the initial files into the `Initializer` class instead of being standalone functions
- using the constants from `constants.dart` to reference the filenames that need to be created

`session.dart`
- new class that will handle session management on the user's computer
- a session will be at minimum 30 minutes (or whatever is defined in `constants.dart`), as the user sends events, the session will end 30 minutes after the latest event sent

`user_property.dart`
- this class will hold all the metadata around a user and will manage getting the latest session id from the `Session` class (extends the session or creates a new session accordingly)

### CODE REVIEW

**Comment by eliasyishak** [comment]
I have also implemented a `constants.dart` file to define all the constant values that get used within the package such as the tools message we will show the user, the duration for a session, etc.

**Comment by eliasyishak** [comment]
General updates for this PR:

`analytics.dart`
- added platform as an arg for the class
- using the constants.dart file to store constant values that are getting passed 
- added a close method which will close the http client in GAClient
- added the sendEvent method to take in eventName and eventData and use GAClient to send data
- instead of default values in the AnalyticsImpl class, we will use the factory constructor in the abstract class to add the default values (allows for easier testing)
- using the GAClient class within the AnalyticsImpl class to send data
- creating a UserProperty instance in the constructor for AnalyticsImpl to store metadata around the user

`config_handler.dart`
- using `package:clock` for date stamps which allows for easier testing when checking session length
- using the `constants.dart` file to store the directory and file names for the files generated

`ga_client.dart`
- created this class to send events to Google Analytics to abstract away creating the HttpClient from the main `Analytics` class

`initializer.dart`
- moving the functions that created the initial files into the `Initializer` class instead of being standalone functions
- using the constants from `constants.dart` to reference the filenames that need to be created

`session.dart`
- new class that will handle session management on the user's computer
- a session will be at minimum 30 minutes (or whatever is defined in `constants.dart`), as the user sends events, the session will end 30 minutes after the latest event sent

`user_property.dart`
- this class will hold all the metadata around a user and will manage getting the latest session id from the `Session` class (extends the session or creates a new session accordingly)

File: example/dash_analytics_example.dart
Comment (eliasyishak) [inline example/dash_analytics_example.dart]: Make this an enum to make sure we have no variations in the strings

File: example/dash_analytics_example.dart
Comment (eliasyishak) [inline example/dash_analytics_example.dart]: Look into Dart's enhanced enums for keeping event names

Comment (christopherfujino) [reply example/dash_analytics_example.dart]: https://dart.dev/guides/language/language-tour#enumerated-types

File: lib/src/analytics.dart
Comment (christopherfujino) [inline lib/src/analytics.dart]: I think this can/should be const
```suggestion
        fs: const LocalFileSystem(),
```

File: lib/src/config_handler.dart
Comment (christopherfujino) [inline lib/src/config_handler.dart:59]: Should this be `'.dart'`?

File: example/dash_analytics_example.dart
Comment (eliasyishak) [reply example/dash_analytics_example.dart]: Thank you! This is perfect, snippet of sample code

```dart
enum Events {
  // List all of the events here
  runTime('run_time'),
  hotReloadTime('hot_reload_time');

  // The positional arg for the default constructor are the labels we have
  // defined in the values above
  const Events(this.label);
  final String label;
}

void main() {
  const event = Events.runTime;
  print(event.label); // Returns 'run_time'
}
```

File: lib/src/config_handler.dart
Comment (eliasyishak) [reply lib/src/config_handler.dart:59]: I wasn't sure on the naming for the constants which is why i created the `constants.dart` file so we can change on the fly.

The PDD did mention a `.dart-tool` directory though and the design doc seems to suggest something else...

Comment (christopherfujino) [reply lib/src/config_handler.dart:59]: I'm not too opinionated on this. I just suggested `.dart` because I already have a user-home level `.dart` but not `.dart-tool`, and I thought `.dart-tool` was specifically for projects.

File: example/dash_analytics_example.dart
Comment (christopherfujino) [inline example/dash_analytics_example.dart]: package consumers generally shouldn't import files under `lib/src` (there's an analyzer lint for this, but apparently it's not part of package:lints?) (note, it's ok for tests to directly import any file). Instead, can you make `package:dash_analytics/dash_analytics.dart` export everything that users would need to directly reference?

Comment (christopherfujino) [reply example/dash_analytics_example.dart]: This is an example of dart file at the top-level lib directory that exists to control what from lib/src will be exported to end users: https://github.com/flutter/flutter/blob/master/packages/flutter/lib/foundation.dart

Comment (eliasyishak) [reply example/dash_analytics_example.dart]: Oh whoops, I completely forgot to replace that – the auto import is too good 😄

File: example/dash_analytics_example.dart
Comment (christopherfujino) [inline example/dash_analytics_example.dart]: This shouldn't be hard-coded, right?

File: lib/src/analytics.dart
Comment (christopherfujino) [inline lib/src/analytics.dart]: Provide explicit types

```suggestion
  Map<String, String> get userPropertyMap;
```

File: lib/src/analytics.dart
Comment (christopherfujino) [inline lib/src/analytics.dart]: Should this be `Future<http.Response?> sendEvent(...)`? Also, whose responsibility should it be to check the status code of the response?

File: lib/src/analytics.dart
Comment (christopherfujino) [inline lib/src/analytics.dart:252]: I know it's not part of this PR, but I'm thinking `setTelemetry` should also be asynchronous, since we're both reading and writing to/from disk, which could be slow.

File: lib/src/analytics.dart
Comment (christopherfujino) [inline lib/src/analytics.dart]: Since this is a test helper, it might make sense to make some of these non-required and provide fallbacks. For instance, if the test doesn't need to verify what was written to disk, this factory could construct its own `MemoryFileSystem()`. Likewise for most (all?) the rest of these fields. This would make an easier API for writing tests, where you only need to provide the fields you want to test.

File: lib/src/analytics.dart
Comment (christopherfujino) [inline lib/src/analytics.dart]: I don't think we should require callers to provide the `Directory homeDirectory` for two reasons:

1. we are relying on the clients to reproduce the logic to determine a user's home directory, which is complicated, and if two clients implement this differently, those two clients won't share the config file
1. the interface `Directory` is from package:file, and would require callers to do something like:

```dart
final homePath = determineUserHomePath();
final homeDir = (const LocalFileSystem()).file(homePath);
```

I think instead package:dash_analytics should take care of both of those steps.

It would make sense for the test class to have an optional override with a `Directory` object.

File: lib/src/analytics.dart
Comment (christopherfujino) [inline lib/src/analytics.dart:138]: can this be private?

File: lib/src/analytics.dart
Comment (christopherfujino) [reply lib/src/analytics.dart]: Can you also enable the `always_specify_types` and `always_declare_return_types` lints in analysis_options.yaml?

File: test/dash_analytics_test.dart
Comment (christopherfujino) [inline test/dash_analytics_test.dart:50]: I think two things are being conflated here:

1. client tests; and
2. package:dash_analytics' tests

This `TestAnalytics` class you've written makes sense for 1, however it is not sufficient for 2 since it shares no code with the real implementation. Essentially, `AnalyticsImpl` is untested.

It would make sense to have a few tests of `Analytics.test`, but essentially you should be testing `AnalyticsImpl` here.

File: lib/src/config_handler.dart
Comment (christopherfujino) [inline lib/src/config_handler.dart:38]: Can you do something like:

```dart
class ConfigHandler {
  ConfigHandler({
    @visibleForTesting
    Clock? clockOverride,
    // etc
  }) : _clock = clockOverride ?? clock {
    // etc
  }
  
  final Clock _clock;
}
```

This way test-owned clocks can be provided.

File: lib/src/config_handler.dart
Comment (christopherfujino) [inline lib/src/config_handler.dart]: You can avoid the `!` operator if you rewrite the first statement in this function as:

```dart
final ToolInfo? toolInfo = parsedTools[tool];
if (toolInfo == null) {
  return;
}
```

And then these lines can be (via type promotion):

```dart
    // Update the [ToolInfo] object for the current tool
    toolInfo.lastRun = clock.now();
    toolInfo.versionNumber = newVersionNumber;
```

File: test/dash_analytics_test.dart
Comment (christopherfujino) [inline test/dash_analytics_test.dart:5]: This (and every other file in this repo) needs to have a LICENSE comment. This also needs a top level LICENSE file.

File: lib/src/constants.dart
Comment (christopherfujino) [inline lib/src/constants.dart]: can this be `dash-analytics.config`, so it's obvious to someone doing `ls .dart-tool` what it is?

File: lib/src/constants.dart
Comment (christopherfujino) [inline lib/src/constants.dart]: can we namespace this, like `dash-analytics-session.json`, or something similar?

File: lib/src/enums.dart
Comment (christopherfujino) [inline lib/src/enums.dart]: nit
```suggestion
  final String description;
```

File: lib/src/enums.dart
Comment (christopherfujino) [inline lib/src/enums.dart:16]: How about we add:

```dart
final String toolOwner;
```

Which, for `hot_reload_time` would be `flutter_tools`.

File: lib/src/enums.dart
Comment (christopherfujino) [inline lib/src/enums.dart]:
```suggestion
enum DevicePlatform {
```

File: lib/src/enums.dart
Comment (christopherfujino) [inline lib/src/enums.dart]: I think this should be the name of the operating system:

```suggestion
  macos('macOS'),
```

Unless there is some other constraining factor?

File: lib/src/enums.dart
Comment (christopherfujino) [reply lib/src/enums.dart]: Where and for what purpose is this used? Are we not currently sending more detailed operating system information with each event?

File: lib/src/ga_client.dart
Comment (christopherfujino) [inline lib/src/ga_client.dart:22]: should we allow tests to provide their own (fake) client?

File: lib/src/initializer.dart
Comment (christopherfujino) [inline lib/src/initializer.dart:43]: As with `config_handler.dart`, can we allow for direct injecting test-controlled clocks?

File: lib/src/session.dart
Comment (christopherfujino) [inline lib/src/session.dart]: avoid `late` except in tests when they are initialized by `setUp()`. One trick to allow a field to be non-nullable when it has to be calculated from a function call is to create a factory that makes the necessary calls (here calls to `.readAsStringSync()` and `jsonDecode`, then returns the result of a call to a private constructor that then initializes the non-nullable fields (e.g. `Session._({this._sessionId, this._lastPing, etc...})`).

File: lib/src/session.dart
Comment (christopherfujino) [inline lib/src/session.dart:67]: Again, consider allowing tests to provide their own clock.

**Review by christopherfujino — CHANGES_REQUESTED** [review]

File: lib/src/analytics.dart
Comment (eliasyishak) [reply lib/src/analytics.dart:138]: I left this one as not private because I thought it might be helpful for an IDE to have access to this possibly? It can provide visibility into what kind of user metadata we are sending. What do you think about exposing that kind of information?

File: test/dash_analytics_test.dart
Comment (eliasyishak) [reply test/dash_analytics_test.dart:50]: I thought by extending the default `AnalyticsImpl` class for `TestAnalytics` and just overriding the `sendEvent` method I was testing the package. Everything else is shared between the two classes... should `TestAnalytics` instead be implementing the abstract class?

File: lib/src/enums.dart
Comment (eliasyishak) [reply lib/src/enums.dart]: I'm indifferent to what we call them, I just asked JaYoung what they called them currently in GA3 and continued. I can relabel them though

File: example/dash_analytics_example.dart
Comment (eliasyishak) [reply example/dash_analytics_example.dart]: I had originally just hard coded this in because it was for the example code, ideally each client would have their own logic for determining the platform and they would pass it in

File: lib/src/analytics.dart
Comment (eliasyishak) [reply lib/src/analytics.dart]: I originally left this as `void` because I didn't want the tools using this package to have to resolve any non-2xx status codes if the request was unsuccessful; instead it would silently fail

I think we may need a discussion what we should do here

File: lib/src/config_handler.dart
Comment (eliasyishak) [reply lib/src/config_handler.dart:38]: This was initially how I did this but realized that this would required having to pass the clock from the top level `Analytics` class to every class it initializes.

Instead I read that `package:clock` avoids this for testing by using the `withClock(Clock clock, T Function() callback)` method. This allows us to pass a fake clock as the first positional parameter and it will essentially do what you have outlined but through the use of zones. I referred to the article below

https://iiro.dev/controlling-time-with-package-clock/

File: lib/src/enums.dart
Comment (eliasyishak) [reply lib/src/enums.dart]: This is being used by the tools implementing this package. The idea was to have each client tool determine what their OS is, then match it against one of these enum values. Once they have assigned an enum value, we will have all relevant information around the description, tool owner, and anything else we want to associate with OS's

It is better than having each client send a string representation of the OS which could vary (ie. macos, macOS, etc.)

File: lib/src/ga_client.dart
Comment (eliasyishak) [reply lib/src/ga_client.dart:22]: This could be useful, but I'm not sure how helpful it will be because the GA servers always send 2xx status codes, even if the data is malformed. Additionally, the server doesn't send back any payload. Open to discuss this further

File: lib/src/session.dart
Comment (eliasyishak) [reply lib/src/session.dart]: This was a great idea, thank you!

File: example/dash_analytics_example.dart
Comment (eliasyishak) [inline example/dash_analytics_example.dart]: @christopherfujino do you suggest that this method that resolves the user's home directory be moved into the package?

File: lib/src/analytics.dart
Comment (christopherfujino) [reply lib/src/analytics.dart]: You make a good point I hadn't considered that it may not be actionable :). On the other hand, why not return the response? The caller can always ignore it.

File: lib/src/config_handler.dart
Comment (eliasyishak) [reply lib/src/config_handler.dart:59]: That is a good point, I will relabel this constant

File: test/dash_analytics_test.dart
Comment (christopherfujino) [reply test/dash_analytics_test.dart:50]: Oops, you're right, I read this wrong.

However, instead of overriding sendEvent, what about if in tests you instead use a `FakeGAClient` instance, that records the http requests it *would* have sent?

File: lib/src/analytics.dart
Comment (christopherfujino) [reply lib/src/analytics.dart:138]: SGTM

File: lib/src/config_handler.dart
Comment (christopherfujino) [reply lib/src/config_handler.dart:38]: I'll leave this one up to you as I don't think there are too many tests clients would write with a fake clock, but I'll point out that you're direct injecting everything else, and using `withClock()` is essentially the opposite direction of https://github.com/flutter/flutter/issues/47161

File: lib/src/enums.dart
Comment (christopherfujino) [reply lib/src/enums.dart]: Two things:

1. Why not have this package detect the operating system?
2. Ahh, I thought that we sent the OS version with each event, but I was wrong.

File: lib/src/ga_client.dart
Comment (christopherfujino) [reply lib/src/ga_client.dart:22]: Ahh, I didn't realize they always return 2xx, lol. I think there's still value in providing a fake client as it means in tests you can verify the request looks correct.

File: example/dash_analytics_example.dart
Comment (christopherfujino) [reply example/dash_analytics_example.dart]: I think that would be a good idea

File: lib/src/enums.dart
Comment (eliasyishak) [reply lib/src/enums.dart]: Good idea, I've moved the logic into the package now

File: lib/src/analytics.dart
Comment (eliasyishak) [reply lib/src/analytics.dart]: That was the initial design I had actually but I wanted to remove the default values so I know exactly what the test environment is using. For example, I need access to the file system from the tests to be able to access the files to check for their existence

```dart
fs = MemoryFileSystem.test();
home = fs.directory(homeDirName);

// The 3 files that should have been generated
clientIdFile = home
    .childDirectory(kDartToolDirectoryName)
    .childFile(kClientIdFileName);
sessionFile =
    home.childDirectory(kDartToolDirectoryName).childFile(kSessionFileName);
configFile =
    home.childDirectory(kDartToolDirectoryName).childFile(kConfigFileName);
});

// Checks below need access to the file system
test('Initializer properly sets up on first run', () {
  expect(clientIdFile.existsSync(), true,
      reason: 'The $kClientIdFileName file was not found');
  expect(sessionFile.existsSync(), true,
      reason: 'The $kSessionFileName file was not found');
  expect(configFile.existsSync(), true,
      reason: 'The $kConfigFileName was not found');
  expect(dartToolDirectory.listSync().length, equals(3),
      reason:
          'There should only be 3 files in the $kDartToolDirectoryName directory');
  expect(analytics.shouldShowMessage, true,
      reason: 'For the first run, analytics should default to being enabled');
});
```

File: lib/src/analytics.dart
Comment (eliasyishak) [reply lib/src/analytics.dart]: That's fair, including that in the next push

File: lib/src/analytics.dart
Comment (eliasyishak) [reply lib/src/analytics.dart:252]: To make this method async, do I only make the `setTelemetry` method async and leave the other methods within `ConfigHandler` as they are (ie. keep i/o related methods like `readAsStringSync()`, `lastModifiedSync()`, etc.) or do those need to also change to their async counterparts?

Is it enough to do this?
```dart

/// NEW UPDATE
Future<void> setTelemetry(bool reportingBool) async {
  _configHandler.setTelemetry(reportingBool);
}
```

```dart

/// PREVIOUS
void setTelemetry(bool reportingBool) {
  _configHandler.setTelemetry(reportingBool);
}
```

File: test/dash_analytics_test.dart
Comment (eliasyishak) [reply test/dash_analytics_test.dart:50]: Good point, as discussed earlier today, I have captured this as an issue here https://github.com/eliasyishak/dash_analytics/issues/16 so that I can dedicate more time to working on mocking the client

File: lib/src/analytics.dart
Comment (christopherfujino) [reply lib/src/analytics.dart]: By making them optional, tests that need to verify the file system can provide it, and the others don't need to provide it.

File: lib/src/analytics.dart
Comment (christopherfujino) [reply lib/src/analytics.dart:252]: you would need to make `_configHandler.setTelemetry` async, migrating readAsStringSync -> readAsString and writeAsStringSync -> writeAsString (I think lastModifiedSync is probably fast enough that it doesn't need to be async).

Also, did you mean to resolve this conversation?

Comment (eliasyishak) [reply lib/src/analytics.dart:252]: Ah, I marked it as resolved because I thought I figured it out by wrapping the config handler's method to set telemetry status in a future.

If I switch the read and write operations to async, wouldn't I have to `await` for them anyway since I need the contents of the file to toggle telemetry status?

File: lib/src/analytics.dart
Comment (eliasyishak) [reply lib/src/analytics.dart]: Makes sense, I've made them optional with default values for the message and version number, and if the filesystem isn't specified in the constructor, it will fall back on a default instance `MemoryFileSystem.test(style: ...)`

File: lib/src/analytics.dart
Comment (eliasyishak) [reply lib/src/analytics.dart:252]: I have also refactored `setTelemetry` to the below code block
```dart
  @override
  Future<void> setTelemetry(bool reportingBool) {
    return Future<void>(() => _configHandler.setTelemetry(reportingBool));
  }
```

I tested it with the example file and it seems to be running asynchronously

Comment (christopherfujino) [reply lib/src/analytics.dart:252]: > I have also refactored `setTelemetry` to the below code block
> 
> ```dart
>   @override
>   Future<void> setTelemetry(bool reportingBool) {
>     return Future<void>(() => _configHandler.setTelemetry(reportingBool));
>   }
> ```
> 
> I tested it with the example file and it seems to be running asynchronously

This is technically asynchronous, however, the slow parts (reading and writing from/to disk) are still synchronous. If you don't want your PR to be blocked by this, you can assign a TODO to me and I will refactor in another PR.

Comment (eliasyishak) [reply lib/src/analytics.dart:252]: As discussed offline, the following issue https://github.com/eliasyishak/dash_analytics/issues/17 captures the work necessary for making this method async

**Review by christopherfujino — APPROVED** [review]
LGTM once there's a TODO about setTelemetry

**Pull request merged**
